onnxruntime/core/providers/cpu/math/einsum_utils/einsum_typed_compute_processor.h

```diff
@@ -0,0 +1,76 @@
+template <typename T>
+class EinsumTypedComputeProcessor {
+ public:
+  explicit EinsumTypedComputeProcessor(OpKernelContext* context, AllocatorPtr allocator,
+                                       concurrency::ThreadPool* tp,
+                                       EinsumComputePreprocessor& einsum_compute_preprocessor,
+                                       void* cublas_handle, void* cuda_ep)
+      : context_(context),
+        allocator_(allocator),
+        tp_(tp),
+        einsum_compute_preprocessor_(einsum_compute_preprocessor),
+        cublas_handle_(cublas_handle),
+        cuda_ep_(cuda_ep) {}
+
+  // Pass-in device specific functions
+  // (Pass-in CPU implementation or CUDA implementation function depending on the kernel using this class)
+  void SetDeviceHelpers(const EinsumOp::DeviceHelpers::Transpose& device_transpose_func,
+                        const EinsumOp::DeviceHelpers::MatMul<T>& device_matmul_func,
+                        const EinsumOp::DeviceHelpers::ReduceSum<T>& device_reduce_sum_func,
+                        const EinsumOp::DeviceHelpers::DataCopy& device_data_copy_func);
+
+  Status Run();
+
+ private:
+  // Private methods -
+
+  // Processes Einsum operands in a pair-wise fashion
+  // Employs Transpose, ReduceSum, and MatMul under the hood
+  // to achieve MatMul(a, b) and reduces (by summing) along specified axes
+  std::unique_ptr<Tensor> PairwiseOperandProcess(const Tensor& left,
+                                                 const TensorShape& left_shape_override,
+                                                 const Tensor& right,
+                                                 const TensorShape& right_shape_override,
+                                                 const std::vector<int64_t>& reduce_dims,
+                                                 bool is_final_pair);
+
+  // Here we take a "candidate output"(candidate output is a tensor that is a permutation and / or a reshape away from the final output),
+  // and after a few operations to get it to the required output structure, copy it to the op's output
+  // The candidate output might contain dims that may not be part of the op's output (i.e.) the dims will have to be unsqueezed
+  void FinalizeOutput(const Tensor& candidate_output,
+                      const std::vector<int64_t>& ordered_subscript_indices_in_candidate);
+
+  // Private members -
+  OpKernelContext* context_;
+  AllocatorPtr allocator_;
+  concurrency::ThreadPool* tp_;
+  EinsumComputePreprocessor& einsum_compute_preprocessor_;
+
+  EinsumOp::DeviceHelpers::Transpose device_transpose_func_;
+  EinsumOp::DeviceHelpers::MatMul<T> device_matmul_func_;
+  EinsumOp::DeviceHelpers::ReduceSum<T> device_reduce_sum_func_;
+  EinsumOp::DeviceHelpers::DataCopy device_data_copy_func_;
+
+  // CuBLAS handle to be used in case the processing is done on the CUDA EP
+  void* cublas_handle_;
+
+  // CUDA EP - required by ReduceSum CUDA kernel
+  void* cuda_ep_;
+};
```

Comment:
nit: this is specific to CUDA EP impl so wondering if you could just have a void* here and the CUDA impl could store the info it needs in a std::pair or some other struct only it knows about. i.e. can we avoid leaking impl details (that there are 2 handles) from the CUDA EP up to this level.

Reply:
Fixed.